porch/controllers/downstreampackages/pkg/controllers/downstreampackage/downstreampackage_controller.go

```diff
@@ -0,0 +1,366 @@
+				BlockOwnerDeletion: nil,
+			}},
+		},
+		Spec: porchapi.PackageRevisionSpec{
+			PackageName:    dp.Spec.Downstream.Package,
+			WorkspaceName:  porchapi.WorkspaceName(workspaceNamePrefix + "1"),
+			RepositoryName: dp.Spec.Downstream.Repo,
+			Tasks: []porchapi.Task{
+				{
+					Type: porchapi.TaskTypeClone,
+					Clone: &porchapi.PackageCloneTaskSpec{
+						Upstream: porchapi.UpstreamPackage{
+							UpstreamRef: &porchapi.PackageRevisionRef{
+								Name: upstream.Name,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if err := r.Client.Create(ctx, newPR); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *DownstreamPackageReconciler) findAndUpdateExistingRevision(ctx context.Context,
+	dp *api.DownstreamPackage,
+	upstream *porchapi.PackageRevision) (*porchapi.PackageRevision, error) {
+	// First, check if a downstream package exists. If not, just return nil. The
+	// caller will create one.
+	downstream := r.getDownstreamPR(dp)
+	if downstream == nil {
+		return nil, nil
+	}
+
+	// Determine if the downstream package needs to be updated. If not, return
+	// the downstream package as-is.
+	if r.isUpToDate(dp, downstream) {
+		return downstream, nil
+	}
+
+	if downstream.Spec.Lifecycle == porchapi.PackageRevisionLifecyclePublished {
+		var err error
+		downstream, err = r.copyPublished(ctx, downstream, dp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return r.updateDraft(ctx, downstream, upstream)
+}
+
+func (r *DownstreamPackageReconciler) getDownstreamPR(dp *api.DownstreamPackage) *porchapi.PackageRevision {
+	downstream := dp.Spec.Downstream
+
+	var latestPublished *porchapi.PackageRevision
+	// the first package revision number that porch assigns is "v1",
+	// so use v0 as a placeholder for comparison
+	latestVersion := "v0"
+
+	for _, pr := range r.packageRevs.Items {
+		// look for the downstream package in the target repo
+		if pr.Spec.RepositoryName != downstream.Repo ||
+			pr.Spec.PackageName != downstream.Package {
+			continue
+		}
+		// check that the downstream package was created by this DownstreamPackage object
+		owned := false
+
+		// TODO: Implement the "AdoptionPolicy" field, which allows the user to decide if
+		//  the controller should adopt existing package revisions or ignore them. For now,
+		//  we just ignore them.
+		for _, owner := range pr.OwnerReferences {
+			if owner.UID == dp.UID {
+				owned = true
+			}
+		}
+		if !owned {
+			// this downstream package doesn't belong to us
+			continue
+		}
+
+		// Check if this PR is a draft. We should only have one draft created by this controller at a time,
+		// so we can just return it.
+		if pr.Spec.Lifecycle != porchapi.PackageRevisionLifecyclePublished {
+			return &pr
+		} else {
+			latestPublished, latestVersion = compare(&pr, latestPublished, latestVersion)
+		}
+	}
+
+	return latestPublished
+}
+
+func compare(pr, latestPublished *porchapi.PackageRevision, latestVersion string) (*porchapi.PackageRevision, string) {
```

Comment:
Do we need to pass in `latestPublished` and `latestVersion` here? It seems like `latestPublished` is always nil and `latestVersion` is always `v0`.